File: UsingFMDB-Objective-C/UsingFMDB-Objective-C/Book.m
//
//  Book.m
//  UsingFMDB-Objective-C
//
//  Created by akabeko on 2016/12/03.
//  Copyright © 2016年 akabeko. All rights reserved.
//

#import "Book.h"

NSInteger const kBookIdNone = 0;

@interface Book ()

@property (nonatomic, readwrite) NSInteger bookId;
@property (nonatomic, readwrite) NSString *author;
@property (nonatomic, readwrite) NSString *title;
@property (nonatomic, readwrite) NSDate *releaseDate;

@end

@implementation Book

/**
 * Create a book instance.
 *
 * @param bookId      Identifier.
 * @param author      Author.
 * @param title       Title.
 * @param releaseDate Release date.
 *
 * @return Instance.
 */
+ (instancetype)bookWithId:(NSInteger)bookId author:(NSString *)author title:(NSString *)title releaseDate:(NSDate *)releaseDate {
    return [[Book alloc] init:bookId author:author title:title releaseDate:releaseDate];
}

#pragma mark - Private

/**
 * Initialize this instance.
 *
 * @param bookId      Identifier.
 * @param author      Author.
 * @param title       Title.
 * @param releaseDate Release date.
 *
 * @return Initialized instance.
 */
- (instancetype)init:(NSInteger)bookId author:(NSString *)author title:(NSString *)title releaseDate:(NSDate *)releaseDate {
    self = [super init];
    if (self) {
        self.bookId      = bookId;
        self.title       = title;
        self.author      = author;
        self.releaseDate = releaseDate;
    }

    return self;
}

@end

File: UsingFMDB-Swift/UsingFMDB-Swift/Book.swift
//
//  Book.swift
//  UsingFMDB-Swift
//
//  Created by akabeko on 2016/12/17.
//  Copyright © 2016年 akabeko. All rights reserved.
//

import UIKit

/// Represents a book.
class Book: NSObject {
    /// Invalid book identifier.
    static let BookIdNone = 0

    /// Identifier.
    private(set) var bookId: Int

    /// Title.
    private(set) var title: String

    /// Author.
    private(set) var author: String

    /// Release date.
    private(set) var releaseDate: Date

    /// Initialize the instance.
    ///
    /// - Parameters:
    ///   - bookId:      Identifier
    ///   - author:      Author.
    ///   - title:       Title.
    ///   - releaseDate: Release Date
    init(bookId: Int, author: String, title: String, releaseDate: Date) {
        self.bookId      = bookId
        self.author      = author
        self.title       = title
        self.releaseDate = releaseDate
    }
}
